Adds NLP queries to search

urbancvek opened the pull request:
This closes #12. This PR allows user to use NLP queries in search by writing queries like this:

`facebook after "10 days ago" before "yesterday"`
`stackoverflow react-native after "last year"`

Date parsing is supported by https://github.com/wanasit/chrono.

It also fills out the time filter fields on overview page.

oliversauter commented:
Great stuff :) will soon test it out. 

Just a small remark: I would change the query so it includes a ":" and remove the space after after/before. (`after:"10 days ago"`) 
This way we can make sure that it belongs together in the query. People should ideally also be able to use " to search for a fix term `Climate change` instead of `climate` and `change` 

Also we could think about what is a better wording instead of before and after, maybe using `from` and `to` (also faster to type)

Just an idea, what do you think about it?

urbancvek commented:
The way that it's implemented now it looks for exactly `after "something something"` it won't match if it doesn't look exactly like that. So yeah I could add ":" or maybe we can set it to match many characters (":", " ", "="). I used space because I thought it was faster to type but can easily change it.

About "from" and "to" I also think this may be better since it looks more like a sentence someone would say and not a specific query language so it's easier to remember.

oliversauter commented:
It just came to my mind, that using the : is also more conform with common ways of querying. For example on google you would type in "keyword site:guardian.co.uk" to filter for that page. Would it also be possible to remove the " for single words? like "keyword to:yesterday" 

Actually now I see that "before" and "after" may be a better choice of words. It is not flowing really well like that. "keyword after:yesterday" is better.

poltak commented:
@urbancvek How are you going with this? The main functionality seems to be working alright :) Found what I think is a little bug in the overview UI code changes, so I'll comment on that part.

I agree with replacing the spaces with `:` in the pattern. Spaces always seem ambiguous to me (like I assume 2+ conseq. spaces acts like 1, but then the regex pattern needs to change slightly). Also would prefer to stick with `before` and `after` keywords; `from` and `to` work great together, but can lose meaning if used separately in this context.

@oliversauter RE: 
> People should ideally also be able to use " to search for a fix term Climate change instead of climate and change

I think that's a little out-of-scope for this PR/issue. It goes back to the `search/find-pages` module's `db.search()` call, which may or may not handle those double quotation marks (I'm not entirely sure as I haven't touched the search logic). But from what I understand of @urbancvek's code now, if you have two similar queries differentiated by double quotation marks like that:

1. `climate change after "10 days ago" before "yesterday"` 
2. `"climate change" after "10 days ago" before "yesterday"`

1 will extract the text `'climate change '` __(single quotes added for reading clarity; note that extra final space - may or may not be intended)__ and pass that to the `search` module as the query.

2 will extract the text `'"climate change" '` and pass that along as the query. Hence it's up to the `search` module how to interpret those double quotes, if that makes sense.

poltak commented on src/overview/actions.js:
```diff
@@ -31,8 +31,25 @@ export function init() {
+        query = extractedQuery || query
+
+        if (oldStartDate !== newStartDate) dispatch(setStartDate({startDate}))
+        if (oldEndDate !== newEndDate) dispatch(setEndDate({endDate}))
```
At the moment these two lines seems to make it impossible to cancel out NLP time queries in the overview. Like I mean if I first type `stackoverflow after "100 mins ago"`, it sets the `startDate` fine, but then if you backspace so the query is just `stackoverflow`, the `startDate` remains unchanged and continues to affect the search results. It should be cleared instead, as it's no longer present in the user's query. Let me know if that example doesn't make sense.

urbancvek replied:
Example makes total sense. This should definitely be changed. There's also a problem for what we should do when a user changes "before" and "after" time filters by using the time pickers. Don't know if we should change the query or what to do... If we don't sync the change to the query the next character that is typed would override date pickers.

oliversauter replied:
> There's also a problem for what we should do when a user changes "before" and "after" time filters by using the time pickers. 

You mean after the user types in the query and then switches to the overview and uses the time pickers there? In this case, there should not be the query in the addressbar anymore anyhow. It would be a url with querystrings.

urbancvek replied:
Ok great was just a misunderstanding. I thought we should also use this NLP stuff in search field on overview. No problem than.

poltak commented on src/util/nlp-time-filter.js:
```diff
@@ -0,0 +1,34 @@
+        matchedBefore ? matchedBefore.index : query.length,
+        matchedAfter ? matchedAfter.index : query.length,
+    )
+    const extractedQuery = query.substring(0, firstIndexOfTimeFilter)
```
May or may not be intended, but using `firstIndexOfTimeFilter` here will include the space between the actual query term and the first NLP time filter literal. Like `facebook after "10 days ago"` will search using the query "facebook " rather than "facebook"

urbancvek replied:
Yes this string could and probably should be trimmed.

poltak commented on src/overview/actions.js:
```diff
@@ -31,8 +31,25 @@ export function init() {
+
+        const startDate = newStartDate || oldStartDate
+        const endDate = newEndDate || oldEndDate
+        query = extractedQuery || query
```
As far as I understand, `extractTimeFiltersFromQuery` will always return a defined `extractedQuery`, hence this line doesn't really need to exist. `query` is just used as input to `extractTimeFiltersFromQuery`, and also means you can make the `let` on L34 a `const` as there's no need to reassign any variables. (tell me if I'm wrong) Nothing major, just something that may easily be overlooked.

urbancvek commented:
I've replaced " " with ":" in regex. I've also removed time filters from search field in overview as it's not in scope of this PR.

When omnibar submits it now sets params in url: query, startDate, endDate. Previously it only set "q". This "q" was a mapping to key in a redux store named "query". I've now renamed it as I think it makes much more sense now that we are 3 params.

Url also syncs on all changes made in a picker. This means you can reload a page and pickers get filled in if they were previously entered :)

poltak commented:
Cool, had a play and it seems to work all good now with the colon instead of space delimiter. And the URL query syncing/populating the overview dates works well!

The addition of `query-string` package and how it is used (simply used to convert the `startDate`, `endDate`, and `query` query keys to URL query string), could be done with a simple string manipulation function, but it's a tiny, and often-used, package, that cleans things up and affords ease of adding any further query params in the future to the search. So, IMO, I think it's an acceptable addition here.

@oliversauter's original issue #12 still explicitly mentions the search-box in overview, which now this PR now does not support (at the request of @oliversauter), so I think before merging in I'll wait for Oliver to comment on that (either we just treat it as a changed requirement in the issue or follow-up with a later PR to fully address the issue?).

Great stuff @urbancvek!

urbancvek commented:
Yeah I think it's OK to use `query-string` package as we'll probably work some more with urls and it simplifies the code.

oliversauter commented:
@poltak @urbancvek 
> still explicitly mentions the search-box in overview, which now this PR now does not support (at the request of @oliversauter)

Yes indeed, it would be good if the search boxes and or (the time filter boxes) also support this kind of querying. If it is not too much hassle, adding it would be great.

poltak commented:
I think there was some confusion on mine and @urbancvek's part then. @urbancvek originally had support for the overview search input, which worked good apart from a bug I pointed out ([this thread](https://github.com/WorldBrain/WebMemex/pull/23#discussion_r117227373)), then @oliversauter commented and @urbancvek (and admittedly me as well) misunderstood @oliversauter's comment to mean remove the functionality on overview search input.

Reading it again, I understand @oliversauter's (it was actually very clear; we just must of had a lapse of judgement). So @urbancvek, I recommend manually revert the changes in 332fec7aeebb32da3aec13c2217bdce42f79b8c8 relating to the `actions.js` file, then see if you can fix that case I pointed out where the bug is when you backspace but the filters remain. It shouldn't be too hard as you already had it working.

One unrelated thing @urbancvek: the need to revert half that commits brings up a good point: That commit would have been better as two separate commits. 1 to remove the overview action code and 1 to add the small fixes to the query matching logic (see how they're two unrelated tasks). Then you could just run `git revert` on the commit that was later deemed unnecessary to automatically create a new commit reverting those changes. So yeah, just for the future, better to try and keep small, concise commits containing changes that relate only to each other (sometimes it becomes blurry; I usually try to see if I'm writing things like "and" or unrelated sentences in the commit message)

urbancvek commented:
Ok yeah at first I did understand the issue but then had a talk with oliver and thought he meant we should only use this nlp from omnibar and use datepickers for time filter on overview.

If we want nlp on overview then there's a problem with syncing input field with date pickers and the other way around.

1. Let's say someone writes:
`facebook after:"10 days ago"`
We then take this query and set the date pickers to reflect that. We also use this to search the history.

2. User then uses date pickers to change the day two days forward.
This change is reflected in datepickers and used to search the history.

3. User types another character into input field and now the query is:
`facebook p after:"10 days ago"`
We take this change and override the datepickers again to reflect the date that is 10 days before.

So you can see that there's a problem that user can override it's own settings. The way to avoid it would be to take the change in step 2 and convert that back into words and sync the input field with the date picker.

Even if we do that in step 2 and the query now says: (updated two days forward by date picker)
`facebook after:"8 days ago"`
If user then deletes last few characters the query could be
`facebook after:"8 da` (last few chars deleted)
When parsing the query then we of course can't find the date which is then null and we set the datepickers to reflect that.

4. User then uses datepickers to some random date like last month
If we then try to sync the query with this date picker we won't be able to find `after:"..."` with regex because it's not complete so we could end up with something like this.
`facebook after:"8 da after:"last month"`

Don't know if it was clear but would probably be hard to find `after:"8 da` in a string especially because words after `after:"` can literally be anything and there's no order for `before`, `after` and `literal query` part.

What I can think of is we could look for `after:` and then replace everything that follows until `before:` or the end of a query. That I think would work ok.

oliversauter commented:
Sorry for the confusion :)

I think a mix of solution 1 and 2 would be suitable. 
To make it easy, why not just ignore queries like "after:"10 days ago" in the search bar. 
People just be able to type NLP queries into the actual date picker fields. 
Basically the date field would also understand queries like "10 days ago" and turn it into a date that is then reflected in the date picker. As soon as the user switches the focus off the field or presses enter, the conversion happens. 
This would prevent us having to do another kind of sync with fields and ship the feature at least satisfyingly?

urbancvek commented:
Yes this would be very easy to implement. You can already type into date pickers so I just add a conversion which is also already implemented just needs to be called.

oliversauter commented:
great, then lets just do it like that for now.

urbancvek commented:
This now fully implements nlp in date pickers as well. You can still use the old way "18-05-2016" to set the date but it also supports NLP.

I've disabled keyboard navigation inside of date pickers (this meant you could cycle through days by using arrow keys) so that you can use arrow keys to move the cursor in text field.

This last commit isn't the prettiest. The problem is `react-datepicker` doesn't allow to pass some function as props that would listen for key presses of any sort. It also uses `"onkeydown"` event itself and uses a few keys. "Enter" which is important to us is amongst them. It is used to set the date to the day that is currently selected in calendar and can be changed by using the arrow keys (I've disabled this in previous commit). If you don't hijack "Enter" key event from date picker it always sets the date to the one selected which is always today because that's the default.

oliversauter commented:
## Testing results:
#### Bugs
1.  when typing in a number like `10 days/weeks/years ago`, it uses the 10. of the current month, not the 15th what it should be when typing `10 days ago`. If you use numbers higher than 31, it works correctly though. i.e. `50 days ago`

> This last commit isn't the prettiest 

Alternative to react-datepicker?

obsidianart commented on src/overview/components/DateRangeSelection.jsx:
```diff
@@ -1,46 +1,126 @@
+
+    componentDidMount() {
+        this.refs.startDateDatePicker.refs.input.refs.input.addEventListener('keydown', e => this.handleKeydown(e, 'startDate'))
+        this.refs.endDateDatePicker.refs.input.refs.input.addEventListener('keydown', e => this.handleKeydown(e, 'endDate'))
```
direct manipulation of dom element is considered a very bad pattern. If it is done as a shortcut for the time being and there is no better way that's fine.

obsidianart replied:
I understood why you did this. If it is not possible to change the Datepicker behaviour I'd rather wrap it in a component which will do this for you (possibly using a proper way) instead of mixing concerns

urbancvek replied:
Yes these two lines are bad. Do you have any experience what would be the best thing to do. 

https://github.com/Hacker0x01/react-datepicker/blob/master/src/datepicker.jsx#L269 This function in datepicker is the one that's causing the problem. "react-datepicker" also wants to listen to key presses so there's no simple way (as far as I know) to prevent this "onInputKeyDown" function on datepicker from being called. Unless you get to the event before React and stop it from propagating. The worst thing is there's really a simple fix to solve this if there was a way to pass in onKeyDown as props to datepicker component.

Do you know of any other way to get this event before datepicker component? Since we're higher in the tree datepicker component always gets it first. Unless you attach a native listener to this input field what I did.

I can extract this to a separate component but I don't know how much the quality of the code will improve since those two lines would still exist and there would just be a lot of passing the same props.

obsidianart commented on src/overview/components/DateRangeSelection.jsx:
```diff
@@ -1,46 +1,126 @@
+    }
+
+    handleKeydown(event, type) {
+        if (event.key === 'Enter') {
```
this is normally done by wrapping the input in a form so that enter does naturally submit the form.

obsidianart replied:
not sure if you can do it in the datepicker...

urbancvek replied:
Explained in the comment above why this is needed.

obsidianart commented on src/overview/components/DateRangeSelection.jsx:
```diff
@@ -1,46 +1,126 @@
+
+    submitDateChange(type) {
+        if (type === 'startDate') {
+            console.log('hrere')
```
forgotten

obsidianart commented on src/overview/components/DateRangeSelection.jsx:
```diff
@@ -1,46 +1,126 @@
+    submitDateChange(type) {
+        if (type === 'startDate') {
+            console.log('hrere')
+            const date = moment(this.state.startDateText, 'DD-MM-YYYY')
```
I don't fully know the reason behind those lines so I might be wrong, but do you really need 2 different libraries to get a date? Chrono seems to also get some IT formats

urbancvek replied:
"chrono-node" when trying to get IT formats does it like this "MM-DD-YYYY". Everywhere else in this component we use "DD-MM-YYYY" and I thought it was very confusing to use two different formats. Also moment is not going away since it's used everywhere in this component and needed for "react-datepicker" so might as well just use especially if it makes code simpler.

I think it would be very confusing to someone why there are months and dates inverted at least it was to me until I found out.. Also means we would had to set the state in a different format.

obsidianart commented on src/util/nlp-time-filter.js:
```diff
@@ -0,0 +1,36 @@
+import chrono from 'chrono-node'
+
+const BEFORE_REGEX = /before:"(.*?)"/
```
I'm more used to ```after:"([^"]*)"``` but they are probably equivalent

urbancvek replied:
Should I change it? It does the same thing.

obsidianart commented on src/util/nlp-time-filter.js:
```diff
@@ -0,0 +1,36 @@
+
+    let startDate
+    let endDate
+    if (matchedBefore && matchedBefore[1]) {
```
I can't find a scenario in which I have ```matchedBefore``` but not ```matchedBefore[1]```

obsidianart commented on src/util/nlp-time-filter.js:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    // Checks which comes first before/after and based on that decide where query should be sliced
+    const firstIndexOfTimeFilter = Math.min(
```
I wonder if the regex doesn't have a way to to this automatically, something like 
```
extractedQuery.replace(BEFORE_REGEX,'').replace(AFTER_REGEX,'').trim()
```
Would it work?

urbancvek replied:
Thanks for this. It works great and much simpler code!

urbancvek commented:
@oliversauter last commit solves the "10 days ago" being parsed into 10th of the month instead of actually the date that's 10 days ago. The issue was the string was picked by moment instead of chrono and moment would try everything just to parse something out.

About the react-datepicker alternatives. I don't know if we should look at a new datepicker component as this onKeydown prop is the only thing that's missing and we only have 2 bad lines because of it. This we can solve by creating a PR  to react-datepicker that would replace the default onKeydown function. If it's accepted we then only have a simple fix of deleting these two bad lines and pass our function (that handles this "onkeydown" event) as props to datepicker component.

poltak commented:
@urbancvek Yeah, I had a play and tried a few things but there doesn't seem to be any react-way of overriding that event without touching the DOM; it's just a limitation of the lib. Seems to be other people who run into issues relating to this ([#751](https://github.com/Hacker0x01/react-datepicker/issues/751), [#802](https://github.com/Hacker0x01/react-datepicker/issues/802), [#580](https://github.com/Hacker0x01/react-datepicker/issues/580)), but sadly the maintainer hasn't responded. 

I think for now, just leave it this way, as it works alright in our case. Maybe put an inline comment briefly explaining why you need to do it (for future contributors who may get confused by it). If you have time, submit a PR upstream to add an optional `onEnter` prop or something (like described in #580), then we may be able to get rid of those lines sometime in the future if it's ever accepted (but for the sake of this PR, I don't consider that super important).

Also try to stay away from string `ref` assignments because [React considers them legacy now](https://facebook.github.io/react/docs/refs-and-the-dom.html#legacy-api-string-refs) for various reasons and could be removed whenever. Can't change the ones used in the date picker lib though 😛

oliversauter commented:
So I played with it also again. All seems to work! 

If you folks @urbancvek @obsidianart @poltak are fine with it. Let's merge that thang!

poltak commented on src/overview/components/DateRangeSelection.jsx:
```diff
@@ -1,46 +1,125 @@
+        }
+    }
+
+    submitDateChange(type) {
```
This method seems a good candidate for DRYing out, as it's just the same code repeated apart from `state.startDateText`, `props.startDate`, `props.onStartDateChange` (and their `endDate` counterparts). Might as well extract them out, switching on `type` in the method body, and return a function that runs the main logic in here (bound to those vars), if that makes sense? If this method returns a function, the calls don't need to be nested in an arrow function (L90/117)

poltak commented on src/overview/components/DateRangeSelection.jsx:
```diff
@@ -1,46 +1,125 @@
+        return (
+            <div className={styles.dateRangeSelection}>
+                <DatePicker
+                    ref='startDateDatePicker'
```
Make this a simple arrow fn like `ref={ref => { this.blah = ref }}` to avoid the whole string ref pattern, then can update the stuff in `componentDidMount()` from `this.refs.blah` to `this.blah` to get the explicitly assigned ref.

poltak commented:
@urbancvek At a quick glance, things look to be nicer now :) cool stuff! I'll double check everything tomorrow morning and merge in if okay.

Also in the future, on feature branches just try to stick to using implicit merges (`git rebase`) if you need to merge in changes from base branch (`master`, in this case). Generally cleaner to leave explicit merges (`git merge`) for merging in things like PRs back into master, to avoid merge commits everywhere. I'll fix it up when I merge in tomorrow though.

poltak commented:
Alright all the new changes from this PR have been merged into master now. Found there were some earlier merge commits that I didn't see yesterday, which further complicated things and ended up with me rebasing all the new commits manually into master (great fun...). So I'll close this PR, but note that every __new__ commit in this PR in now in master.

oliversauter commented:
Thanks! Great stuff. So this could have prevented with rebasing before making the final PR?

poltak commented:
@oliversauter Yeah, and just generally avoid explicit merges from base branch to feature branches like this. If there's really a need to sync changes from the base branch (which, most of the time, probably is master) use rebase to replay the branch-specific changes on top of whatever master currently is. 

As the master branch is constantly going to be rebased on top of upstream changes, the history is going to keep getting force overwritten resulting in some weird looking changes on PRs (like can be seen in this PR). It's a bit awkward, due to the awkward way position we're in with the two repos, but I think it should be fine if we follow the general rule of: __only ever use explicit merges when merging feature branches back into master.__

Anyone feel free to call me out if you know of a better way.
